Add config file

patxoca opened the pull request:
See issue #16

I've tried to make simple to manage the options while keeping the code minimal. A nice improvement would be creating the config parser dynamically from the schema but I felt like being too much.

I've done some profiling and looking for, reading an parsing the config file does not hurt performance noticeably (milliseconds in an old i5 laptop).

The code requires testing on windows.

patxoca committed:
Prototype for using a config file.

patxoca committed:
Generalize implementation.

patxoca committed:
Move "parse_command_line" function.

patxoca committed:
Code refactor.

patxoca committed:
Replace unserscores by spaces in config file options.

patxoca committed:
Added section "Config file" to the readme.

patxoca committed:
Fix bug with boolean options.

When using bool as the type all non empty string are True.

patxoca committed:
Code documentation.

reidrac commented on z80count/z80count.py:
```diff
@@ -34,6 +37,180 @@
+    try:
+        parser.read(config_file)
+    except configparser.ParsingError:
+        perror("Error parsing config file. Using defaults.")
```
May be provide some pointer on what when wrong?

reidrac commented on z80count/z80count.py:
```diff
@@ -34,6 +37,180 @@
+            v = opt.type(v)
+        except (ValueError, TypeError):
+            perror(
+                "Error parsing config value for '%s'. Using default.",
```
Same, please provide some top on what went wrong and potentially how to fix it.

reidrac commented on z80count/z80count.py:
```diff
@@ -34,6 +37,180 @@
+    if xdg_config_home is None:
+        xdg_config_home = os.path.join(home_dir, ".config")
+
+    candidate = os.path.join(xdg_config_home, "z80count")
```
Use a constant for the configuration file; I prefer it is `z80countrc`.

reidrac commented on z80count/z80count.py:
```diff
@@ -34,6 +37,180 @@
+    if os.path.isfile(candidate):
+        return candidate
+
+    candidate = os.path.join(home_dir, ".z80count")
```
Same, use `"." + DEF_CONF` or similar.

reidrac commented on z80count/z80count.py:
```diff
@@ -34,6 +37,180 @@
+        description='Z80 Cycle Count',
+        epilog="Copyright (C) 2019 Juan J Martinez <jjm@usebox.net>")
+
+    # NOTE: parser arguments *must* provide a default equal to None in
```
You could have used as default the value from the configuration file (that would be default if not conf file was provided).

reidrac commented on README.md:
```diff
@@ -152,6 +152,38 @@ Where A, B, T0 and T1 are:
+- `update`: boolean
+- `use tabs`: boolean
+
+Example:
```
I would include a `z80countrc` as example with the defaults commented. It is easier to copy the file and uncomment/change whatever, helping the README to not have too much information (that perhaps should go into a man page).

reidrac commented on README.md:
```diff
@@ -152,6 +152,38 @@ Where A, B, T0 and T1 are:
+
+- the file given in the environment variable `Z80COUNT_RC`.
+
+- a file `z80count` in the directory given in the environment variable
```
I prefer this to be `z80countrc`.

reidrac reviewed (commented):
A bit cumbersome for such a small tool, but I don't mind the change.

There few things I would improve, specially regarding error reporting (considering this change adds configuration file support, sounds like an important thing).

patxoca committed:
Improve error reporting.

patxoca committed:
Change config file filename.

patxoca committed:
Provide defaults to "parse_command_line".

Rationale for the current implementation: using an explicit merge of
the options provided in the config file and command line options
allows us to swap the order in which "load_config_file" and
"parse_command_line" are called. That way switches provided in the
command line may influence the processing of the config file.

Using meaningful defaults in "parse_command_line" we loose the ability
to detect if a switch has been provided or is using the default,
effectively imposing an order to the calls.

This flexibility is not required for "z80count" so we may loose it in
exchange of a simpler "parse_command_line". Reverting to the current
implementation would be a matter of passing an _all None_ dictionary
to "parse_command_line" and adding back the explicit merge.

patxoca committed:
Rename config option.

Having an option that has positive semantics in the config file and
negative in the command line calls for problems. Convert "update" to
"keep cycles" ("no_update" synonym).

patxoca committed:
Update readme.

patxoca commented:
I've done the requested changes. Anyway, I've been thinking that having approximately 40% of the code for managing the configuration makes little sense.

If you're OK with that I'll withdraw the PR. In the future, if a real need arises, we'll talk again.

reidrac commented:
I agree it is a bit over-engineered, but I'm keeping it. Thanks!

patxoca commented:
A minor detail that does not deserve an issue, in the readme, in the config file example, please remove the final period for `keep cycles` comment. I've missed that one.
